File: appmart/overrides/controllers/salary_structure_assignment.py
# Copyright (c) 2021, The Nexperts Pvt. Ltd. and Contributors
# License: GNU General Public License v3. See license.txt

import frappe
from frappe import _
from frappe.utils import flt


def before_save(doc,method):
    if False and doc.get("base") and doc.get("employee_salary_components"):
        total_amount = flt(0)
        for comp in doc.get("employee_salary_components"):
            if comp.get("type") == "Earning":
                total_amount += comp.get("amount") or 0
        if total_amount > doc.get("base"):
            frappe.throw("<b> Salary Components </b> total cannot be greater than <b> Base </b>")

@frappe.whitelist()
def get_components(doctype='Salary Structure Assignment', txt='', searchfield='name', start=0, page_len=20, filters=None, as_dict=False):
    return frappe.db.get_values("Salary Detail",filters={"parent":filters.get("salary_structure")},fieldname=["salary_component"]) or []

File: appmart/public/js/salary_structure_assignment.js
frappe.ui.form.on('Salary Structure Assignment', {
    refresh:function(frm) {
        frm.set_query('salary_component', 'employee_salary_components', function() {
            if(!frm.doc.salary_structure) {
                frappe.throw(__('Please set Salary Structure'));
            }
            return {
                query: 'appmart.overrides.controllers.salary_structure_assignment.get_components',
                filters: {
                    salary_structure: frm.doc.salary_structure
                }
            };
        });
    }
});